taichi/program/async_engine.cpp

```diff
@@ -206,4 +221,78 @@ bool AsyncEngine::optimize() {
   return modified;
 }
 
+bool AsyncEngine::fuse() {
+  // TODO: improve...
+  bool modified = false;
+  std::unordered_map<SNode *, bool> list_dirty;
+
+  if (false) {
+    // (experimental) print tasks
+    for (int i = 0; i < (int)task_queue.size(); i++) {
+      fmt::print("{}: {}\n", i, task_queue[i].stmt->task_name());
+      irpass::print(task_queue[i].stmt);
+    }
+  }
+
+  for (int i = 0; i < (int)task_queue.size() - 1; i++) {
+    auto task_a = task_queue[i].stmt;
+    auto task_b = task_queue[i + 1].stmt;
+    bool is_same_struct_for = task_a->task_type == OffloadedStmt::struct_for &&
+                              task_b->task_type == OffloadedStmt::struct_for &&
+                              task_a->snode == task_b->snode &&
+                              task_a->block_dim == task_b->block_dim;
+    bool is_same_range_for = task_a->task_type == OffloadedStmt::range_for &&
+                             task_b->task_type == OffloadedStmt::range_for &&
+                             task_a->const_begin && task_b->const_begin &&
+                             task_a->const_end && task_b->const_end &&
+                             task_a->begin_value == task_b->begin_value &&
+                             task_a->end_value == task_b->end_value;
+
+    // We do not fuse serial kernels for now since they can be SNode accessors
+    bool are_both_serial = task_a->task_type == OffloadedStmt::serial &&
+                           task_b->task_type == OffloadedStmt::serial;
+    bool same_kernel = task_queue[i].kernel == task_queue[i + 1].kernel;
```

Comment:
BTW, thanks to #915, I realized that kernel fusion has some preconditions.. For example, i don't think `clear` and `listgen` are easily fusable, because the latter depends on the fact that all threads would observe the result of `clear`.

While I don't think this change would fuse those kernels, I'd like to confirm that: kernel fusion would only be done if the later kernel doesn't rely on the kernel boundary being a barrier?